Fix footnote links broken when pasting from Word

juanruitina opened the pull request:
By adding 'name' to allowed attributes for 'a' elements
Fixes #12784

<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->
When copying text with footnotes, Word links footnote numbers and footnotes to each other with anchors that make use of the `name` attribute, as explained [here](https://stigmortenmyre.no/mso/html/word/wdconfootnotesandendnotes.htm). The file I edited, phrasing-content.js, was removing all attributes that were not in that array, including `name`, so I just added `, 'name'` to it. This solves the issue.

It's maybe worth noting though that the use of the `name` attribute for `a` elements [is obsolete in HTML5](https://html.spec.whatwg.org/multipage/obsolete.html#obsolete-but-conforming-features); use of `id` is recommended instead. [W3C Markup Validator](https://validator.w3.org/) also shows a warning. An alternate solution would therefore be to save the content of the `name` attribute as the element's `id`. Maybe this is just overcomplicating things, though.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
Tested on a WordPress 5.5.3 installation.
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md --> (Not sure about this)

juanruitina committed:
Fix footnote links broken when pasting from Word

By adding 'name' to allowed attributes for 'a' elements
See issue #12784 https://github.com/WordPress/gutenberg/issues/12784#issuecomment-743156872

ellatrix commented:
What's the resulting HTML going into the block? Maybe we should upgrade the HTML to use id instead of name, if that markup is deprecated?

juanruitina committed:
Use ID instead of name for footnote links

name attribute is obsolete in HTML5

juanruitina commented:
I've been playing with it a bit and the way to use `id` instead of `name` is, well, adding `id` instead of `name` to the array 😅 Apparently the code copied from Word has the same value from both, but the id was also getting removed. I think I got confused because TinyMCE seems to strip `id` values, but not `name`.

The only change to the resulting HTML is that it adds the `id` parameter and value as copied from Word. It goes from this:

```html
<!-- wp:paragraph -->
<p>This test has a couple<a href="#_ftn1">[1]</a> of footnotes<a href="#_ftn2">[2]</a>.</p>
<!-- /wp:paragraph -->

<!-- wp:separator -->
<hr class="wp-block-separator"/>
<!-- /wp:separator -->

<!-- wp:paragraph -->
<p><a href="#_ftnref1">[1]</a> This is footnote 1.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p><a href="#_ftnref2">[2]</a> This is footnote 2.</p>
<!-- /wp:paragraph -->
```

to this:

```html
<!-- wp:paragraph -->
<p>This test has a couple<a href="#_ftn1" id="_ftnref1">[1]</a> of footnotes<a href="#_ftn2" id="_ftnref2">[2]</a>.</p>
<!-- /wp:paragraph -->

<!-- wp:separator -->
<hr class="wp-block-separator"/>
<!-- /wp:separator -->

<!-- wp:paragraph -->
<p><a href="#_ftnref1" id="_ftn1">[1]</a> This is footnote 1.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p><a href="#_ftnref2" id="_ftn2">[2]</a> This is footnote 2.</p>
<!-- /wp:paragraph -->
```

I think it is not possible to edit `id`s for inline tags from the visual editor, but I can't think of any potential issues with keeping these ids, particularly when they will only be kept for `a` tags? Would having these attributes only accessible in code be a bad practice?

ellatrix commented:
It's accessible in the HTML view, so it's fine for now.

I do think we should only preserve ids if there's a link for it, otherwise we'll paste a lot of useless ids from all kinds of web pages. Could you create a new filter that removes ids if they don't have a link? This code should be general, it should work for any id.

ellatrix commented on packages/dom/src/phrasing-content.js:
```diff
@@ -20,7 +20,7 @@ const textContentSchema = {
 	del: {},
 	ins: {},
-	a: { attributes: [ 'href', 'target', 'rel' ] },
+	a: { attributes: [ 'href', 'target', 'rel', 'id' ] },
```
If we allow ids for links, we should allow them for all tags, but only the content has a link to it.

juanruitina replied:
@ellatrix One question about this: why do you think it is important to allow 'id' for all elements? The scope of this issue is to fix the broken footnotes (and perhaps endnotes) when pasting from Word, and only anchor elements are involved.

ellatrix replied:
Because generally, any element could be linked to, for example a heading. I agree, let's add this to one element at a time though.

juanruitina committed:
Save name attribute as id and remove

Allow name attribute for a elements before reducing

juanruitina committed:
Keep only a element' ids for footnotes and endnotes

Compatible with Word and LibreOffice

juanruitina committed:
Remove a tag if no link nor id

juanruitina commented:
Will think about how to handle internal links as you suggest. (Sorry for the new commits, pushed them too soon, I'm trying some things. For now at least works for footnotes-endnotes, but will try to do something more "universal")

juanruitina committed:
Add tests for handling Word/LibreOffice footnotes

juanruitina committed:
Add LibreOffice to raw-handling documentation

juanruitina commented:
I've had a look at this again. I've checked footnotes and endnotes from Word and LibreOffice work. I've also added a test for correct handling of IDs and to check obsolete `name` attributes are removed. I've updated the documentation for raw-handling to add LibreOffice. 

@ellatrix, as you suggest, there could be some check for keeping internal links beyond these footnotes and endnotes. I feel however this is beyond the scope of this issue and I understand it would be more complex, something I'd not feel confortable coding myself. Could we perhaps merge these changes and create a different issue for this?

(In this case, there's no need to check the IDs have some link pointing to them because, when copying text that contains references to footnotes and endnotes, both programmes will always append the footnotes and endnotes themselves.)

ellatrix commented on packages/blocks/src/api/raw-handling/phrasing-content-reducer.js:
```diff
@@ -56,5 +56,26 @@ export default function phrasingContentReducer( node, doc ) {
+		// Search for common footnote and endnote id structure (Word, LibreOffice). If doesn't match, remove id attribute.
+		if (
+			node.id &&
+			! /^_?(?:ftn|edn|sdfootnote|sdendnote)/i.test( node.id )
```
Not sure if I was clear before, but I don't agree with this part. Could we check `node.ownerDocument.querySelector('[href=${ node.id }]')` instead so we know if there's actually a link to this element or not? If no link exists to the element with this ID, then we should not keep the ID around.

ellatrix replied:
The philosophy of the paste filters is to keep them as general as possible and not start matching identifiers from specific platforms.

juanruitina replied:
Thank you for clarifying this and for the guidance, I totally get what you mean now. Will try to make it work.
